libsolidity/codegen/ExpressionCompiler.cpp

```diff
@@ -1259,6 +1259,12 @@ bool ExpressionCompiler::visit(FunctionCall const& _functionCall)
 				// function call.
 				solAssert(arguments.size() == 2, "");
 				solAssert(function.kind() == FunctionType::Kind::Require, "");
+				// Check if the function call matches ``require(bool, error)``, and throw an unimplemented error,
+				// as require with custom errors is not supported in legacy codegen.
+				auto const* magicType = dynamic_cast<MagicType const*>(arguments[1]->annotation().type);
+				if (magicType && magicType->kind() == MagicType::Kind::Error)
+					solUnimplemented("Require with a custom error is only available using the via-ir pipeline.");
```

Comment:
Wait, I see in `syntaxTests/errors/require_custom_legacy.sol` that in legacy the unimplemented error is actually caught and presented as a normal error. But then this simple snippet gives me an ICE on the CLI:

`test.sol`
```solidity
error CustomError(uint256, string);

contract C {
    function f(bool condition) external pure {
        require(condition, CustomError(1, "two"));
    }
}
```
```bash
solc test.sol --bin
```
```
Unimplemented feature:
/solidity/libsolidity/codegen/ExpressionCompiler.cpp(1266): Throw in function virtual bool solidity::frontend::ExpressionCompiler::visit(const solidity::frontend::FunctionCall&)
Dynamic exception type: boost::wrapexcept<solidity::langutil::UnimplementedFeatureError>
std::exception::what: Require with a custom error is only available using the via-ir pipeline.
[solidity::util::tag_comment*] = Require with a custom error is only available using the via-ir pipeline.
```

Reply:
soltest catches these and outputs the error nicely - that's why the outputs differ. Honestly I don't even remember what we decided with this; I do remember spending a day or two trying to figure out source locations and why they're lost during exception propagation (the source location *is* present when the unimplemented exception is thrown, but is then lost when the exception is caught). There's Daniel's [branch](https://github.com/ethereum/solidity/compare/develop...legacyUnimplementedSourceLocation), and some of the tests are failing due to expectations not being update, but I can't for the life of me remember what we decided to do with this (this was before Sicily)...

Reply:
I think the last time we talked about it you said that you didn't apply that change because it was not giving you a line number and I thought I convinced you that displaying a proper error even without a line number was still better than nothing.

But anyway, like I said on the chat a moment ago, I think we should address this, but it would be perfectly fine done in a PR of its own. My main concern was being able to test the new behavior on legacy and seems that this part actually does work here. And the problem is really bigger - `require` just happens to be the most common trigger right now - so doing it separately makes sense.